// gatsby-cli/lib/reporter/constants.js
"use strict";

exports.__esModule = true;
exports.ActivityTypes = exports.ActivityStatuses = exports.ActivityLogLevels = exports.LogLevels = exports.Actions = void 0;
const Actions = {
  LogAction: `LOG_ACTION`,
  SetStatus: `SET_STATUS`,
  Log: `LOG`,
  SetLogs: `SET_LOGS`,
  StartActivity: `ACTIVITY_START`,
  EndActivity: `ACTIVITY_END`,
  UpdateActivity: `ACTIVITY_UPDATE`,
  PendingActivity: `ACTIVITY_PENDING`,
  CancelActivity: `ACTIVITY_CANCEL`,
  ActivityErrored: `ACTIVITY_ERRORED`
};
exports.Actions = Actions;
const LogLevels = {
  Debug: `DEBUG`,
  Success: `SUCCESS`,
  Info: `INFO`,
  Warning: `WARNING`,
  Log: `LOG`,
  Error: `ERROR`
};
exports.LogLevels = LogLevels;
const ActivityLogLevels = {
  Success: `ACTIVITY_SUCCESS`,
  Failed: `ACTIVITY_FAILED`,
  Interrupted: `ACTIVITY_INTERRUPTED`
};
exports.ActivityLogLevels = ActivityLogLevels;
const ActivityStatuses = {
  InProgress: `IN_PROGRESS`,
  NotStarted: `NOT_STARTED`,
  Interrupted: `INTERRUPTED`,
  Failed: `FAILED`,
  Success: `SUCCESS`,
  Cancelled: `CANCELLED`
};
exports.ActivityStatuses = ActivityStatuses;
const ActivityTypes = {
  Spinner: `spinner`,
  Hidden: `hidden`,
  Progress: `progress`,
  Pending: `pending`
};
exports.ActivityTypes = ActivityTypes;
