// @typescript-eslint/eslint-plugin/dist/rules/require-await.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const experimental_utils_1 = require("@typescript-eslint/experimental-utils");
const require_await_1 = __importDefault(require("eslint/lib/rules/require-await"));
const tsutils = __importStar(require("tsutils"));
const util = __importStar(require("../util"));
exports.default = util.createRule({
    name: 'require-await',
    meta: {
        type: 'suggestion',
        docs: {
            description: 'Disallow async functions which have no `await` expression',
            category: 'Best Practices',
            recommended: 'error',
            requiresTypeChecking: true,
        },
        schema: require_await_1.default.meta.schema,
        messages: require_await_1.default.meta.messages,
    },
    defaultOptions: [],
    create(context) {
        const rules = require_await_1.default.create(context);
        const parserServices = util.getParserServices(context);
        const checker = parserServices.program.getTypeChecker();
        /**
         * Checks if the node returns a thenable type
         *
         * @param {ASTNode} node - The node to check
         * @returns {boolean}
         */
        function isThenableType(node) {
            const type = checker.getTypeAtLocation(node);
            return tsutils.isThenableType(checker, node, type);
        }
        return {
            'FunctionDeclaration[async = true]': rules.FunctionDeclaration,
            'FunctionExpression[async = true]': rules.FunctionExpression,
            'ArrowFunctionExpression[async = true]'(node) {
                rules.ArrowFunctionExpression(node);
                // If body type is not BlockStatment, we need to check the return type here
                if (node.body.type !== experimental_utils_1.AST_NODE_TYPES.BlockStatement) {
                    const expression = parserServices.esTreeNodeToTSNodeMap.get(node.body);
                    if (expression && isThenableType(expression)) {
                        // tell the base rule to mark the scope as having an await so it ignores it
                        rules.AwaitExpression(node);
                    }
                }
            },
            'FunctionDeclaration[async = true]:exit': rules['FunctionDeclaration:exit'],
            'FunctionExpression[async = true]:exit': rules['FunctionExpression:exit'],
            'ArrowFunctionExpression[async = true]:exit': rules['ArrowFunctionExpression:exit'],
            AwaitExpression: rules.AwaitExpression,
            ForOfStatement: rules.ForOfStatement,
            ReturnStatement(node) {
                const { expression } = parserServices.esTreeNodeToTSNodeMap.get(node);
                if (expression && isThenableType(expression)) {
                    // tell the base rule to mark the scope as having an await so it ignores it
                    rules.AwaitExpression(node);
                }
            },
        };
    },
});
