// react-hotkeys/cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "HotKeys", {
  enumerable: true,
  get: function get() {
    return _HotKeys.default;
  }
});
Object.defineProperty(exports, "GlobalHotKeys", {
  enumerable: true,
  get: function get() {
    return _GlobalHotKeys.default;
  }
});
Object.defineProperty(exports, "IgnoreKeys", {
  enumerable: true,
  get: function get() {
    return _IgnoreKeys.default;
  }
});
Object.defineProperty(exports, "ObserveKeys", {
  enumerable: true,
  get: function get() {
    return _ObserveKeys.default;
  }
});
Object.defineProperty(exports, "withHotKeys", {
  enumerable: true,
  get: function get() {
    return _withHotKeys.default;
  }
});
Object.defineProperty(exports, "withIgnoreKeys", {
  enumerable: true,
  get: function get() {
    return _withIgnoreKeys.default;
  }
});
Object.defineProperty(exports, "withObserveKeys", {
  enumerable: true,
  get: function get() {
    return _withObserveKeys.default;
  }
});
Object.defineProperty(exports, "configure", {
  enumerable: true,
  get: function get() {
    return _configure.default;
  }
});
Object.defineProperty(exports, "getApplicationKeyMap", {
  enumerable: true,
  get: function get() {
    return _getApplicationKeyMap.default;
  }
});

var _HotKeys = _interopRequireDefault(require("./HotKeys"));

var _GlobalHotKeys = _interopRequireDefault(require("./GlobalHotKeys"));

var _IgnoreKeys = _interopRequireDefault(require("./IgnoreKeys"));

var _ObserveKeys = _interopRequireDefault(require("./ObserveKeys"));

var _withHotKeys = _interopRequireDefault(require("./withHotKeys"));

var _withIgnoreKeys = _interopRequireDefault(require("./withIgnoreKeys"));

var _withObserveKeys = _interopRequireDefault(require("./withObserveKeys"));

var _configure = _interopRequireDefault(require("./configure"));

var _getApplicationKeyMap = _interopRequireDefault(require("./getApplicationKeyMap"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
