// gatsby-transformer-sharp/fragments.js
"use strict";

exports.__esModule = true;
exports.GatsbyImageSharpSizes_withWebp_noBase64 = exports.GatsbyImageSharpSizes_noBase64 = exports.GatsbyImageSharpSizes_withWebp_tracedSVG = exports.GatsbyImageSharpSizes_withWebp = exports.GatsbyImageSharpSizes_tracedSVG = exports.GatsbyImageSharpSizes = exports.GatsbyImageSharpResolutions_withWebp_noBase64 = exports.GatsbyImageSharpResolutions_noBase64 = exports.GatsbyImageSharpResolutions_withWebp_tracedSVG = exports.GatsbyImageSharpResolutions_withWebp = exports.GatsbyImageSharpResolutions_tracedSVG = exports.GatsbyImageSharpResolutions = exports.GatsbyImageSharpFluid_withWebp_noBase64 = exports.GatsbyImageSharpFluid_noBase64 = exports.GatsbyImageSharpFluid_withWebp_tracedSVG = exports.GatsbyImageSharpFluid_withWebp = exports.GatsbyImageSharpFluid_tracedSVG = exports.GatsbyImageSharpFluid = exports.GatsbyImageSharpFixed_withWebp_noBase64 = exports.GatsbyImageSharpFixed_noBase64 = exports.GatsbyImageSharpFixed_withWebp_tracedSVG = exports.GatsbyImageSharpFixed_withWebp = exports.GatsbyImageSharpFixed_tracedSVG = exports.GatsbyImageSharpFixed = void 0;

var _gatsby = require("gatsby");

/* eslint-disable */
const GatsbyImageSharpFixed = _gatsby.graphql`
  fragment GatsbyImageSharpFixed on ImageSharpFixed {
    base64
    width
    height
    src
    srcSet
  }
`;
/**
 * Traced SVG fixed images
 * @type {Fragment}
 */

exports.GatsbyImageSharpFixed = GatsbyImageSharpFixed;
const GatsbyImageSharpFixed_tracedSVG = _gatsby.graphql`
  fragment GatsbyImageSharpFixed_tracedSVG on ImageSharpFixed {
    tracedSVG
    width
    height
    src
    srcSet
  }
`;
/**
 * Images using Webp for fixed images
 * @type {Fragment}
 */

exports.GatsbyImageSharpFixed_tracedSVG = GatsbyImageSharpFixed_tracedSVG;
const GatsbyImageSharpFixed_withWebp = _gatsby.graphql`
  fragment GatsbyImageSharpFixed_withWebp on ImageSharpFixed {
    base64
    width
    height
    src
    srcSet
    srcWebp
    srcSetWebp
  }
`;
/**
 * Traced SVG images using Webp for fixed images
 * @type {Fragment}
 */

exports.GatsbyImageSharpFixed_withWebp = GatsbyImageSharpFixed_withWebp;
const GatsbyImageSharpFixed_withWebp_tracedSVG = _gatsby.graphql`
  fragment GatsbyImageSharpFixed_withWebp_tracedSVG on ImageSharpFixed {
    tracedSVG
    width
    height
    src
    srcSet
    srcWebp
    srcSetWebp
  }
`;
/**
 * Fixed images without the blurred base64 image
 * @type {Fragment}
 */

exports.GatsbyImageSharpFixed_withWebp_tracedSVG = GatsbyImageSharpFixed_withWebp_tracedSVG;
const GatsbyImageSharpFixed_noBase64 = _gatsby.graphql`
  fragment GatsbyImageSharpFixed_noBase64 on ImageSharpFixed {
    width
    height
    src
    srcSet
  }
`;
/**
 * Fixed images without the blurred base64 image preferring Webp
 * @type {Fragment}
 */

exports.GatsbyImageSharpFixed_noBase64 = GatsbyImageSharpFixed_noBase64;
const GatsbyImageSharpFixed_withWebp_noBase64 = _gatsby.graphql`
  fragment GatsbyImageSharpFixed_withWebp_noBase64 on ImageSharpFixed {
    width
    height
    src
    srcSet
    srcWebp
    srcSetWebp
  }
`;
/**
 * The simplest set of fields for fluid images
 * @type {Fragment}
 */

exports.GatsbyImageSharpFixed_withWebp_noBase64 = GatsbyImageSharpFixed_withWebp_noBase64;
const GatsbyImageSharpFluid = _gatsby.graphql`
  fragment GatsbyImageSharpFluid on ImageSharpFluid {
    base64
    aspectRatio
    src
    srcSet
    sizes
  }
`;
/**
 * Traced SVG fluid images
 * @type {Fragment}
 */

exports.GatsbyImageSharpFluid = GatsbyImageSharpFluid;
const GatsbyImageSharpFluid_tracedSVG = _gatsby.graphql`
  fragment GatsbyImageSharpFluid_tracedSVG on ImageSharpFluid {
    tracedSVG
    aspectRatio
    src
    srcSet
    sizes
  }
`;
/**
 * Fluid images that prefer Webp
 * @type {Fragment}
 */

exports.GatsbyImageSharpFluid_tracedSVG = GatsbyImageSharpFluid_tracedSVG;
const GatsbyImageSharpFluid_withWebp = _gatsby.graphql`
  fragment GatsbyImageSharpFluid_withWebp on ImageSharpFluid {
    base64
    aspectRatio
    src
    srcSet
    srcWebp
    srcSetWebp
    sizes
  }
`;
/**
 * Traced SVG fluid images that prefer Webp
 * @type {Fragment}
 */

exports.GatsbyImageSharpFluid_withWebp = GatsbyImageSharpFluid_withWebp;
const GatsbyImageSharpFluid_withWebp_tracedSVG = _gatsby.graphql`
  fragment GatsbyImageSharpFluid_withWebp_tracedSVG on ImageSharpFluid {
    tracedSVG
    aspectRatio
    src
    srcSet
    srcWebp
    srcSetWebp
    sizes
  }
`;
/**
 * Traced SVG fluid images without the blurred base64 image
 * @type {Fragment}
 */

exports.GatsbyImageSharpFluid_withWebp_tracedSVG = GatsbyImageSharpFluid_withWebp_tracedSVG;
const GatsbyImageSharpFluid_noBase64 = _gatsby.graphql`
  fragment GatsbyImageSharpFluid_noBase64 on ImageSharpFluid {
    aspectRatio
    src
    srcSet
    sizes
  }
`;
/**
 * Traced SVG fluid images without the blurred base64 image that prefer Webp
 * @type {Fragment}
 */

exports.GatsbyImageSharpFluid_noBase64 = GatsbyImageSharpFluid_noBase64;
const GatsbyImageSharpFluid_withWebp_noBase64 = _gatsby.graphql`
  fragment GatsbyImageSharpFluid_withWebp_noBase64 on ImageSharpFluid {
    aspectRatio
    src
    srcSet
    srcWebp
    srcSetWebp
    sizes
  }
`; // TODO: in v3 remove these legacy fragments

exports.GatsbyImageSharpFluid_withWebp_noBase64 = GatsbyImageSharpFluid_withWebp_noBase64;
const GatsbyImageSharpResolutions = _gatsby.graphql`
  fragment GatsbyImageSharpResolutions on ImageSharpResolutions {
    base64
    width
    height
    src
    srcSet
  }
`;
exports.GatsbyImageSharpResolutions = GatsbyImageSharpResolutions;
const GatsbyImageSharpResolutions_tracedSVG = _gatsby.graphql`
  fragment GatsbyImageSharpResolutions_tracedSVG on ImageSharpResolutions {
    tracedSVG
    width
    height
    src
    srcSet
  }
`;
exports.GatsbyImageSharpResolutions_tracedSVG = GatsbyImageSharpResolutions_tracedSVG;
const GatsbyImageSharpResolutions_withWebp = _gatsby.graphql`
  fragment GatsbyImageSharpResolutions_withWebp on ImageSharpResolutions {
    base64
    width
    height
    src
    srcSet
    srcWebp
    srcSetWebp
  }
`;
exports.GatsbyImageSharpResolutions_withWebp = GatsbyImageSharpResolutions_withWebp;
const GatsbyImageSharpResolutions_withWebp_tracedSVG = _gatsby.graphql`
  fragment GatsbyImageSharpResolutions_withWebp_tracedSVG on ImageSharpResolutions {
    tracedSVG
    width
    height
    src
    srcSet
    srcWebp
    srcSetWebp
  }
`;
exports.GatsbyImageSharpResolutions_withWebp_tracedSVG = GatsbyImageSharpResolutions_withWebp_tracedSVG;
const GatsbyImageSharpResolutions_noBase64 = _gatsby.graphql`
  fragment GatsbyImageSharpResolutions_noBase64 on ImageSharpResolutions {
    width
    height
    src
    srcSet
  }
`;
exports.GatsbyImageSharpResolutions_noBase64 = GatsbyImageSharpResolutions_noBase64;
const GatsbyImageSharpResolutions_withWebp_noBase64 = _gatsby.graphql`
  fragment GatsbyImageSharpResolutions_withWebp_noBase64 on ImageSharpResolutions {
    width
    height
    src
    srcSet
    srcWebp
    srcSetWebp
  }
`;
exports.GatsbyImageSharpResolutions_withWebp_noBase64 = GatsbyImageSharpResolutions_withWebp_noBase64;
const GatsbyImageSharpSizes = _gatsby.graphql`
  fragment GatsbyImageSharpSizes on ImageSharpSizes {
    base64
    aspectRatio
    src
    srcSet
    sizes
  }
`;
exports.GatsbyImageSharpSizes = GatsbyImageSharpSizes;
const GatsbyImageSharpSizes_tracedSVG = _gatsby.graphql`
  fragment GatsbyImageSharpSizes_tracedSVG on ImageSharpSizes {
    tracedSVG
    aspectRatio
    src
    srcSet
    sizes
  }
`;
exports.GatsbyImageSharpSizes_tracedSVG = GatsbyImageSharpSizes_tracedSVG;
const GatsbyImageSharpSizes_withWebp = _gatsby.graphql`
  fragment GatsbyImageSharpSizes_withWebp on ImageSharpSizes {
    base64
    aspectRatio
    src
    srcSet
    srcWebp
    srcSetWebp
    sizes
  }
`;
exports.GatsbyImageSharpSizes_withWebp = GatsbyImageSharpSizes_withWebp;
const GatsbyImageSharpSizes_withWebp_tracedSVG = _gatsby.graphql`
  fragment GatsbyImageSharpSizes_withWebp_tracedSVG on ImageSharpSizes {
    tracedSVG
    aspectRatio
    src
    srcSet
    srcWebp
    srcSetWebp
    sizes
  }
`;
exports.GatsbyImageSharpSizes_withWebp_tracedSVG = GatsbyImageSharpSizes_withWebp_tracedSVG;
const GatsbyImageSharpSizes_noBase64 = _gatsby.graphql`
  fragment GatsbyImageSharpSizes_noBase64 on ImageSharpSizes {
    aspectRatio
    src
    srcSet
    sizes
  }
`;
exports.GatsbyImageSharpSizes_noBase64 = GatsbyImageSharpSizes_noBase64;
const GatsbyImageSharpSizes_withWebp_noBase64 = _gatsby.graphql`
  fragment GatsbyImageSharpSizes_withWebp_noBase64 on ImageSharpSizes {
    aspectRatio
    src
    srcSet
    srcWebp
    srcSetWebp
    sizes
  }
`;
exports.GatsbyImageSharpSizes_withWebp_noBase64 = GatsbyImageSharpSizes_withWebp_noBase64;
