// rtl-css-js/dist/cjs/core.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var __chunk_1 = require('./chunk-2697f590.js');



exports.propertyValueConverters = __chunk_1.propertyValueConverters;
exports.propertiesToConvert = __chunk_1.propertiesToConvert;
exports.propsToIgnore = __chunk_1.propsToIgnore;
exports.valuesToConvert = __chunk_1.valuesToConvert;
exports.convert = __chunk_1.convert;
exports.convertProperty = __chunk_1.convertProperty;
exports.getPropertyDoppelganger = __chunk_1.getPropertyDoppelganger;
exports.getValueDoppelganger = __chunk_1.getValueDoppelganger;
exports.arrayToObject = __chunk_1.arrayToObject;
exports.calculateNewBackgroundPosition = __chunk_1.calculateNewBackgroundPosition;
exports.calculateNewTranslate = __chunk_1.flipTransformSign;
exports.flipTransformSign = __chunk_1.flipTransformSign;
exports.flipSign = __chunk_1.flipSign;
exports.handleQuartetValues = __chunk_1.handleQuartetValues;
exports.includes = __chunk_1.includes;
exports.isBoolean = __chunk_1.isBoolean;
exports.isFunction = __chunk_1.isFunction;
exports.isNullOrUndefined = __chunk_1.isNullOrUndefined;
exports.isNumber = __chunk_1.isNumber;
exports.isObject = __chunk_1.isObject;
exports.isString = __chunk_1.isString;
exports.getValuesAsList = __chunk_1.getValuesAsList;
