// @storybook/components/dist/placeholder/placeholder.stories.js
"use strict";

require("core-js/modules/es.object.define-property");

require("core-js/modules/es.object.get-own-property-descriptor");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.twoChildren = exports.singleChild = exports["default"] = void 0;

var _react = _interopRequireWildcard(require("react"));

var _placeholder = require("./placeholder");

var _link = require("../typography/link/link");

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj["default"] = obj; return newObj; } }

var _default = {
  component: _placeholder.Placeholder,
  title: 'Basics|Placeholder'
};
exports["default"] = _default;

var _ref =
/*#__PURE__*/
_react["default"].createElement(_placeholder.Placeholder, null, "This is a placeholder with single child, it's bolded");

var singleChild = function singleChild() {
  return _ref;
};

exports.singleChild = singleChild;
singleChild.displayName = "singleChild";

var _ref2 =
/*#__PURE__*/
_react["default"].createElement(_placeholder.Placeholder, null, _react["default"].createElement(_react.Fragment, {
  key: "title"
}, "This has two children, the first bold"), _react["default"].createElement(_react.Fragment, {
  key: "desc"
}, "The second normal weight. Here's a", ' ', _react["default"].createElement(_link.Link, {
  href: "https://storybook.js.org",
  secondary: true,
  cancel: false
}, "link")));

var twoChildren = function twoChildren() {
  return _ref2;
};

exports.twoChildren = twoChildren;
twoChildren.displayName = "twoChildren";
