// react-focus-lock/dist/cjs/MoveFocusInside.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _focusLock = _interopRequireWildcard(require("focus-lock"));

var _util = require("./util");

var MoveFocusInside =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(MoveFocusInside, _Component);

  function MoveFocusInside() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, MoveFocusInside);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(MoveFocusInside)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)((0, _assertThisInitialized2.default)(_this)), "setObserveNode", function (ref) {
      _this.observed = ref;

      _this.moveFocus();
    });
    return _this;
  }

  (0, _createClass2.default)(MoveFocusInside, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this.moveFocus();
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps) {
      if (prevProps.disabled && !this.props.disabled) {
        this.moveFocus();
      }
    }
  }, {
    key: "moveFocus",
    value: function moveFocus() {
      var observed = this.observed;

      if (!this.props.disabled && observed) {
        if (!(0, _focusLock.focusInside)(observed)) {
          (0, _focusLock.default)(observed, null);
        }
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          children = _this$props.children,
          disabled = _this$props.disabled,
          className = _this$props.className;
      return _react.default.createElement("div", (0, _extends2.default)({}, (0, _util.inlineProp)(_focusLock.constants.FOCUS_AUTO, !disabled), {
        ref: this.setObserveNode,
        className: className
      }), children);
    }
  }]);
  return MoveFocusInside;
}(_react.Component);

exports.default = MoveFocusInside;
(0, _defineProperty2.default)(MoveFocusInside, "defaultProps", {
  disabled: false,
  className: undefined
});
MoveFocusInside.propTypes = process.env.NODE_ENV !== "production" ? {
  children: _propTypes.default.node.isRequired,
  disabled: _propTypes.default.bool,
  className: _propTypes.default.string
} : {};
