// @antv/gatsby-theme-antv/lib/components/Footer.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importStar(require("react"));
const rc_footer_1 = __importDefault(require("rc-footer"));
const react_i18next_1 = require("react-i18next");
const antd_1 = require("antd");
const react_use_1 = require("react-use");
const getProducts_1 = require("./getProducts");
const Header_1 = require("./Header");
const Footer_module_less_1 = __importDefault(require("./Footer.module.less"));
require("rc-footer/assets/index.less");
exports.OLD_SITE_DOMAIN = 'https://antv-2018.alipay.com';
const Footer = ({ columns, bottom, theme = 'dark', language, rootDomain = '', chinaMirrorNotice = true, githubUrl = '', }) => {
    const { t, i18n } = react_i18next_1.useTranslation();
    const lang = language || i18n.language;
    const products = getProducts_1.getProducts({
        t,
        language: lang,
        rootDomain,
    });
    const isWide = react_use_1.useMedia('(min-width: 767.99px)', true);
    react_1.useEffect(() => {
        const timeout = setTimeout(() => {
            if (!chinaMirrorNotice ||
                lang !== 'zh' ||
                window.location.host.includes('chartcube') ||
                window.location.host.includes('gitee.io') ||
                localStorage.getItem('china-mirror-no-more-hint') ||
                !isWide) {
                return;
            }
            antd_1.notification.info({
                key: 'china-mirror',
                icon: (react_1.default.createElement("img", { width: 32, src: "https://gw.alipayobjects.com/zos/antfincdn/FLrTNDvlna/antv.png", alt: "antv logo" })),
                message: 'å›½å†…é•œåƒæç¤º ðŸ‡¨ðŸ‡³',
                description: 'AntV ç³»åˆ—ç½‘ç«™éƒ¨ç½²åœ¨ gh-pages ä¸Šï¼Œè‹¥è®¿é—®é€Ÿåº¦ä¸ä½³ï¼Œå¯ä»¥å‰å¾€å›½å†…é•œåƒç«™ç‚¹ã€‚ä½ ä¹Ÿå¯ä»¥åœ¨é¡¶éƒ¨å¯¼èˆªæ‰¾åˆ°é•œåƒé“¾æŽ¥ã€‚',
                duration: 0,
                top: 64,
                btn: (react_1.default.createElement(react_1.default.Fragment, null,
                    react_1.default.createElement(antd_1.Button, { type: "primary", size: "small", style: { marginRight: 8 }, onClick: () => {
                            Header_1.redirectToChinaMirror(githubUrl);
                        } },
                        react_1.default.createElement(antd_1.Icon, { type: "thunderbolt" }),
                        "\u524D\u5F80\u56FD\u5185\u955C\u50CF"),
                    react_1.default.createElement(antd_1.Button, { size: "small", onClick: () => {
                            localStorage.setItem('china-mirror-no-more-hint', Date.now().toString());
                            antd_1.notification.close('china-mirror');
                        } }, "\u4E0D\u518D\u63D0\u9192"))),
                onClose: () => {
                    localStorage.setItem('china-mirror-no-more-hint', Date.now().toString());
                },
            });
        }, 5000);
        return () => {
            clearTimeout(timeout);
        };
    });
    const more = {
        icon: (react_1.default.createElement("img", { src: "https://gw.alipayobjects.com/zos/rmsportal/nBVXkrFdWHxbZlmMbsaH.svg", alt: "more products" })),
        title: t('æ›´å¤šäº§å“'),
        items: [
            {
                icon: (react_1.default.createElement("img", { src: "https://gw.alipayobjects.com/zos/rmsportal/KDpgvguMpGfqaHPjicRK.svg", alt: "Ant Design" })),
                title: 'Ant Design',
                url: 'https://ant.design',
                description: t('ä¼ä¸šçº§ UI è®¾è®¡è¯­è¨€'),
                openExternal: true,
            },
            {
                icon: (react_1.default.createElement("img", { src: "https://gw.alipayobjects.com/zos/rmsportal/XuVpGqBFxXplzvLjJBZB.svg", alt: "yuque" })),
                title: t('è¯­é›€'),
                url: 'https://yuque.com',
                description: t('çŸ¥è¯†åˆ›ä½œä¸Žåˆ†äº«å·¥å…·'),
                openExternal: true,
            },
            {
                icon: (react_1.default.createElement("img", { src: "https://gw.alipayobjects.com/zos/antfincdn/sAEs8aHCnd/yunfengdie.png", alt: "yunfengdie" })),
                title: t('äº‘å‡¤è¶'),
                url: 'https://yunfengdie.com',
                description: t('ä¸­å°å»ºç«™å¹³å°'),
                openExternal: true,
            },
            {
                icon: (react_1.default.createElement("img", { src: "https://gw.alipayobjects.com/zos/antfincdn/v2%24rh7lqpu/82f338dd-b0a6-41bc-9a86-58aaa9df217b.png", alt: "Egg" })),
                title: 'Egg',
                url: 'https://eggjs.org',
                description: t('ä¼ä¸šçº§ Node å¼€å‘æ¡†æž¶'),
                openExternal: true,
            },
            {
                icon: (react_1.default.createElement("img", { src: "https://gw.alipayobjects.com/zos/rmsportal/DMDOlAUhmktLyEODCMBR.ico", alt: "kitchen" })),
                title: 'Kitchen',
                description: t('Sketch å·¥å…·é›†'),
                url: 'https://kitchen.alipay.com',
                openExternal: true,
            },
            {
                icon: (react_1.default.createElement("img", { src: "https://gw.alipayobjects.com/zos/rmsportal/nBVXkrFdWHxbZlmMbsaH.svg", alt: "xtech" })),
                title: t('èš‚èšä½“éªŒç§‘æŠ€'),
                url: 'https://xtech.antfin.com/',
                openExternal: true,
            },
        ],
    };
    const defaultColumns = products
        .filter(product => product.category !== 'ecology')
        .map(product => ({
        title: (react_1.default.createElement("span", null,
            product.title,
            react_1.default.createElement("span", { className: Footer_module_less_1.default.description }, product.slogan))),
        items: product.links,
    }));
    return (react_1.default.createElement(rc_footer_1.default, { maxColumnsPerRow: 4, theme: theme, columns: columns || [...defaultColumns, more], className: Footer_module_less_1.default.footer, bottom: bottom || (react_1.default.createElement("div", { className: Footer_module_less_1.default.bottom },
            react_1.default.createElement("div", null,
                react_1.default.createElement("a", { href: "https://weibo.com/antv2017", target: "_blank", rel: "noopener noreferrer" },
                    react_1.default.createElement(antd_1.Icon, { type: "weibo" })),
                react_1.default.createElement("a", { href: "https://zhuanlan.zhihu.com/aiux-antv", target: "_blank", rel: "noopener noreferrer" },
                    react_1.default.createElement(antd_1.Icon, { type: "zhihu" })),
                react_1.default.createElement("a", { href: "https://github.com/antvis", target: "_blank", rel: "noopener noreferrer" },
                    react_1.default.createElement(antd_1.Icon, { type: "github" })),
                react_1.default.createElement("a", { href: `${rootDomain}/${lang}/about` }, t('å…³äºŽæˆ‘ä»¬')),
                react_1.default.createElement("a", { href: exports.OLD_SITE_DOMAIN, target: "_blank", rel: "noopener noreferrer" }, t('è¿”å›žæ—§ç‰ˆ'))),
            react_1.default.createElement("div", null,
                "\u00A9 ",
                new Date().getFullYear(),
                " Made with \u2764 by",
                ' ',
                react_1.default.createElement("a", { href: "https://xtech.antfin.com/" }, "XTech")))) }));
};
exports.default = Footer;
