// inversify-logging/dist/loggers/NullLogger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var NullLogger = (function () {
    function NullLogger() {
    }
    NullLogger.prototype.debug = function () {
        var messages = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            messages[_i] = arguments[_i];
        }
    };
    NullLogger.prototype.info = function () {
        var messages = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            messages[_i] = arguments[_i];
        }
    };
    NullLogger.prototype.warning = function () {
        var messages = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            messages[_i] = arguments[_i];
        }
    };
    NullLogger.prototype.error = function () {
        var errors = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            errors[_i] = arguments[_i];
        }
    };
    NullLogger.prototype.setLogLevel = function () {
    };
    NullLogger.prototype.createChildLogger = function (context) {
        return new NullLogger();
    };
    return NullLogger;
}());
exports.NullLogger = NullLogger;
exports.default = new NullLogger();
